feat(mol): add support for fixed-size Union

phroi opened the pull request:
Improve codecs by supporting fixed-size unions

For nervosnetwork/molecule#99, related to #33

Hanssen0 commented on packages/core/src/molecule/codec.ts:
```diff
@@ -444,8 +445,22 @@ export function union<T extends Record<string, CodecLike<any, any>>>(
   const keys = Object.keys(codecLayout);
+  const values = Object.values(codecLayout);
+  let byteLength = values[0]?.byteLength;
+  for (const { byteLength: l } of values.slice(1)) {
```
If the first byteLength is undefined, we can return early to avoid extra checks. Consider to remove the `silce(1)` here.

Hanssen0 commented on packages/core/src/molecule/codec.ts:
```diff
@@ -444,8 +445,22 @@ export function union<T extends Record<string, CodecLike<any, any>>>(
   fields?: Record<keyof T, number | undefined | null>,
 ): Codec<UnionEncodable<T>, UnionDecoded<T>> {
   const keys = Object.keys(codecLayout);
+  const values = Object.values(codecLayout);
```
Consider using `Object.entries()` for performance reasons.

Hanssen0 commented on packages/core/src/molecule/signed.ts:
```diff
@@ -0,0 +1,211 @@
+ * @param littleEndian - Indicates whether to use little-endian byte order.
+ * @returns A codec for 16-bit signed integers.
+ */
+function int16Number(littleEndian: boolean): Codec<NumLike, number> {
```
I prefer to have sth like `intNumber(byteLength, littleEndian)` to reduce duplicate code, though that means we might not be able to use `DataView`.

`ccc.numToByte` added support for negative numbers. Maybe that helps

https://github.com/ckb-devrel/ccc/pull/177

phroi commented:
Thank you for your comments, but I'd rather abandon this Pull Request Draft, as its code holds no value:

1. [`ccc.numToByte` support for negative integers](https://github.com/ckb-devrel/ccc/pull/177) is exactly what's needed here, but also keep in mind that it's a foot-gun if used incorrectly: `IntN` will need to carefully check integer bounds to make sure that no two numbers map to the same representation, for example `2^(8*bytes)  - 1` and `-1`. [For example said checks are completely missing from uint](https://github.com/ckb-devrel/ccc/issues/178)

2.  Fixed Union has been proposed, but [its current form has a different implementation than the one implemeted in this PR](https://github.com/nervosnetwork/molecule/issues/99#issuecomment-2723542625) (thanks to implicit padding feature)

Love & Peace, Phroi

phroi commented:
Reopening this PR: here we'll exclusively just focus on the union codec, adding signed integer codecs is the scope for another PR

phroi commented:
@Hanssen0 I fixed the small performance issues you noted and added some TypeDocs.

Now comes the most difficult part: [should we allow for implicit padding or not?](https://github.com/nervosnetwork/molecule/issues/99#issuecomment-2724182007) What's your take?

In my opinion I don't really think we should cause it feels kind of a breaking change and makes this a bit more out of reach.

Phroi %59

Hanssen0 commented:
Sorry misclicked!
